### PR TITLE
Functional MLIR/CUDA backend, works for small/medium size matrices. 

### PR DESCRIPTION
**Pull request by chudur-budur**
1. Works for small/medium size matrices, fails on some (not all) large matrices.
2. Need to revisit the `gpu.host_register` dialect issue. Throws `'cuMemHostRegister(ptr, sizeBytes, 0)' failed with 'CUDA_ERROR_HOST_MEMORY_ALREADY_REGISTERED'`, but matmul results are correct.
3. Runs on V100,

### CODE REVIEW

File: matmul/CMakeLists.txt [hunk 1]
```diff
@@ -290,13 +323,23 @@ foreach(BACKEND ${BACKENDS})
...
+      # target_include_directories(${MATMUL} PRIVATE /usr/local/cuda-11.0/targets/x86_64-linux/include)
+      # target_compile_options(${MATMUL} PRIVATE -lcuda) # We might need this in future
     elseif(${BACKEND} STREQUAL "nodai")
       target_compile_definitions(${MATMUL} PRIVATE MLIR)
     elseif(${BACKEND} STREQUAL "naive")
       target_compile_definitions(${MATMUL} PRIVATE NAIVE)
       target_compile_options(${MATMUL} PRIVATE -O3)
     endif()
-    target_link_libraries(${MATMUL} PRIVATE m)
+    if(NOT ${BACKEND} STREQUAL "mlircuda")
```
Comment (nick-furry) [inline matmul/CMakeLists.txt:340]: Comments on why the linking is mutually exclusive would be helpful to understand

File: README.md [hunk 2]
```diff
@@ -26,75 +26,120 @@ Note: 8GB Mac Mini runs roughly 25% slower than the 16GB version on other tests.
 ### Installation
 Clone the repo along with submodules.
 
-```
+```bash
```
Comment (nick-furry) [inline README.md:29]: This probably is a residual from editing, consider removing all instances

File: matmul/main.cc [hunk 3]
```diff
@@ -22,6 +22,10 @@
 #include "cuda_runtime.h"
 #include "cublas_v2.h"
 #include <sstream>
+// TODO: fix the cuMemHostRegister/Unregister problem
+// #elif defined(MLIR_CUDA)
```
Comment (nick-furry) [inline matmul/main.cc:26]: Is it `if defined(MLIR_CUDA)`

File: matmul/main.cc [hunk 4]
```diff
@@ -155,6 +159,16 @@ void init_matrix(float *a, int nrows, int ncols) {
   }
 }
 
+// A print function for sanity check
```
Comment (nick-furry) [inline matmul/main.cc:162]: L#391-408 contains a snippet for sanity check, best to reuse it

**Review by nick-furry — CHANGES_REQUESTED** [review]

File: matmul/CMakeLists.txt [hunk 5]
```diff
@@ -210,10 +240,13 @@ foreach(BACKEND ${BACKENDS})
     list(GET SIZES 2 K)
 
     if(${BACKEND} STREQUAL "mlir")
-      compile_mlir(${MATMUL} ${M} ${N} ${K} OFF)
+      compile_mlir(${MATMUL} ${M} ${N} ${K} OFF OFF)
```
Comment (yzhang93) [inline matmul/CMakeLists.txt:243]: why there are two off?

File: matmul/CMakeLists.txt [hunk 6]
```diff
@@ -210,10 +240,13 @@ foreach(BACKEND ${BACKENDS})
...
       list(APPEND SRC ${CMAKE_BINARY_DIR}/mlir-objs/${MATMUL}.o)
     elseif(${BACKEND} STREQUAL "nodai")
-      compile_mlir(${MATMUL} ${M} ${N} ${K} ON)
+      compile_mlir(${MATMUL} ${M} ${N} ${K} ON OFF)
```
Comment (yzhang93) [inline matmul/CMakeLists.txt:246]: what are these ON OFF and OFF ON?

File: matmul/main.cc [hunk 7]
```diff
@@ -22,6 +22,10 @@
 #include "cuda_runtime.h"
 #include "cublas_v2.h"
 #include <sstream>
+// TODO: fix the cuMemHostRegister/Unregister problem
+// #elif defined(MLIR_CUDA)
+// #include "cuda_runtime_api.h"
+// #include "cuda.h"
```
Comment (yzhang93) [inline matmul/main.cc:28]: Useless codes? You need to remove all dead codes.

File: matmul/main.cc [hunk 8]
```diff
@@ -342,6 +385,9 @@ int main(int argc, char **argv) {
   CHECK_CUBLAS(cublasGetVector(MDIM * NDIM, sizeof(float), CC, 1, C, 1));
 #endif
 
+// print result here
```
Comment (yzhang93) [inline matmul/main.cc:388]: Remove dead codes.

File: matmul/main.cc [hunk 9]
```diff
@@ -214,6 +230,20 @@ int main(int argc, char **argv) {
...
+// #if defined(MLIR_CUDA)
+//   float *DA, *DB, *DC;
+//   cudaMalloc((void **)&DA, MDIM * KDIM * sizeof(float));
+//   cudaMalloc((void **)&DB, MDIM * KDIM * sizeof(float));
+//   cudaMalloc((void **)&DC, MDIM * KDIM * sizeof(float));
+//
+//   cudaMemcpy(DA, A, MDIM * KDIM * sizeof(float), cudaMemcpyHostToDevice);
+//   cudaMemcpy(DB, B, MDIM * KDIM * sizeof(float), cudaMemcpyHostToDevice);
+//   cudaMemcpy(DC, C, MDIM * KDIM * sizeof(float), cudaMemcpyHostToDevice);
+// #endif
```
Comment (yzhang93) [inline matmul/main.cc:245]: Will you have another patch soon with this problem solved? The codes are not very readable if you put so many TODOs here and there with codes commented out.

File: matmul/CMakeLists.txt (hunk 5 above)
Comment (chudur-budur) [reply matmul/CMakeLists.txt:243]: The first switch is to turn on/off nodai and the second one is for mlircua. So we need two of them.

File: matmul/CMakeLists.txt (hunk 6 above)
Comment (chudur-budur) [reply matmul/CMakeLists.txt:246]: > what are these ON OFF and OFF ON?

The first switch is to turn on/off nodai and the second one is for mlircua. So we need two of them.

File: matmul/CMakeLists.txt (hunk 5 above)
Comment (chudur-budur) [reply matmul/CMakeLists.txt:243]: > why there are two off?

The first switch is to turn on/off nodai and the second one is for mlircua. So we need two of them.

**Review by nick-furry — APPROVED** [review]

**Pull request merged**